### ISSUE TITLE
⚙ [Support Request] big compilation time (linux - gcc/clang)

### ISSUE THREAD

**Issue by schnek** [AscEmu/AscEmu#1111]
**Description**: 

The problem is that Linux (gcc) has a very long compilation time. Time is approximately equal to building a project without pch. (on the windows)

**Steps to reproduce**: 
1. start compiling on linux
2. after project_logon , you can go to sleep 😁

**Expansion version**: 

    Specify version - is required:
    - Patch 1.12.1 - Classic
    - Patch 2.4.3 - TBC
    - Patch 3.3.5 - WoTLK
    - Patch 4.3.4 - Cataclysm
    - Patch 5.4.8 - Mop

**AscEmu hash/commit**: 

latest dev/master

**Operating system**: 

    Specify the operating system - is required:
    - Platform: Linux

**Comment by schnek** [comment]
https://stackoverflow.com/questions/3935183/forward-declaration-include-on-top-of-declaration-include-classfwd-h-class-h

now I'm thinking of doing something like this 

![image](https://github.com/AscEmu/AscEmu/assets/1216225/c93e2d77-4b80-473a-8503-6dc18f96b74a)

Chat+Fwd.h
Data+Fwd.h
...
roughly looks like this

**Comment by Zyres** [comment]
Not really a solution to solve the problem.
Why is PCH not working on linux? it is generated by cmake for it and should keep compilation time low since it is only compiled once. Maybe we should more headers to pch. did you checked the time difference on linux with pch and without pch?

**Comment by schnek** [comment]
starting from the "world" project, the compilation speed drops and then the project builds for a very long time.
does not depend on pch and non-pch on/off.

**Comment by Appled** [comment]
I can confirm this. CMake use pch option doesn't seem to affect compilation time at all.

**Comment by Zyres** [comment]
The cmake option should work (works for other projects as intended). We increased the count of headerfiles a lot over the past years. There has to be a problem with linux ignoring pch. Digging into it right now.

**Comment by schnek** [comment]
maybe it was removed when remove cotire..

**Comment by schnek** [comment]
![image](https://github.com/AscEmu/AscEmu/assets/1216225/7adcbacb-3805-4a59-a8ee-c4c0344a6c78)
on Windows it turned out to achieve the same build speed. settings are the same with "enabled pch". switch /Yu to -> /Yc.
/MP is blocked.

**Comment by schnek** [comment]
clang.. same..

**Comment by Zyres** [comment]
The compile time should now be shorter. Can you check that?

**Comment by Winfidonarleyan** [comment]
I'll check tomorrow.

**Comment by Winfidonarleyan** [comment]
`8.2m` with 15 threads 126500FK
GCC 13.2.1

**Comment by schnek** [comment]
40m+ with 4 threads GCC 12.2.0


Clang 14.0.6 - failed
```

In file included from <built-in>:441:
In file included from /home/ascemu/Development/AscEmu/src/world/CMakeFiles/world.dir/cmake_pch.hxx:5:
In file included from /home/ascemu/Development/AscEmu/src/world/pchWorld.hpp:8:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/vector:63:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_uninitialized.h:64:
In file included from /usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/ext/alloc_traits.h:34:
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/alloc_traits.h:518:4: error: no matching function for call to 'construct_at'
          std::construct_at(__p, std::forward<_Args>(__args)...);
          ^~~~~~~~~~~~~~~~~
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/vector.tcc:117:21: note: in instantiation of function template specialization 'std::allocator_traits<std::allocator<AIMessage>>::construct<AIMessage, unsigned int &, bool &>' requested here
            _Alloc_traits::construct(this->_M_impl, this->_M_impl._M_finish,
                           ^
/home/ascemu/Development/AscEmu/src/world/Server/Script/AIUtils.hpp:208:27: note: in instantiation of function template specialization 'std::vector<AIMessage>::emplace_back<unsigned int &, bool &>' requested here
        mAIMessages[slot].emplace_back(textId, useTarget);
                          ^
/usr/bin/../lib/gcc/x86_64-linux-gnu/12/../../../../include/c++/12/bits/stl_construct.h:94:5: note: candidate template ignored: substitution failure [with _Tp = AIMessage, _Args = <unsigned int &, bool &>]: no matching constructor for initialization of 'AIMessage'
    construct_at(_Tp* __location, _Args&&... __args)
    ^
In file included from /home/ascemu/Development/AscEmu/src/world/Chat/Commands/DebugCommands.cpp:21:
In file included from /home/ascemu/Development/AscEmu/src/world/Server/Script/CreatureAIScript.hpp:18:
In file included from /home/ascemu/Development/AscEmu/src/world/Server/Script/CreatureAIFunctionScheduler.hpp:10:

```

**Comment by schnek** [comment]
I rechecked Cmake and in many examples the implementation is exactly the same as it is in AE. 
https://github.com/nikitabuyevich/cmake-pch-boost-example/blob/master/pkg/int/foobar/CMakeLists.txt#L13

**Comment by schnek** [comment]
reassembling the code. maybe need to do something with the linker

```
# required for linker.
set(SRC_BUILD_WORLD_PVP_SCRIPTS
    ../../world/Server/ServerState.cpp
```

**Comment by Zyres** [comment]
Just to check in on this issue. On Win it is 2:36 right now

**Comment by schnek** [comment]
Do you fast computer. I have average of 8m now.


<img width="988" height="171" alt="Image" src="https://github.com/user-attachments/assets/7cfc2167-8fbb-4378-acf5-3fb235adca45" />

windows ~20 m

<img width="626" height="43" alt="Image" src="https://github.com/user-attachments/assets/0f8eb1fd-941c-4ae9-82c6-9516569e443f" />

macos ~20m

<img width="619" height="38" alt="Image" src="https://github.com/user-attachments/assets/4df8a9eb-a31e-45e7-a858-a82fe2e7f734" />

linux clang ~50m

<img width="620" height="38" alt="Image" src="https://github.com/user-attachments/assets/ded98266-334e-4f15-92b3-283c0b5fbb9e" />

linux gcc ~50m

I think we still need to push 😅

**Comment by schnek** [comment]
Yeah, maybe, but it's also possible that it rebuilds the project from scratch every time without using any cache..😁👌

**Comment by schnek** [comment]
@Appled  If you think the compilation speed is more or less on par with “Windows,” then just close the ticket.
As I mentioned earlier, you need to add various “cache” options during compilation, and this will speed up the build.

**Comment by Appled** [comment]
> If you think the compilation speed is more or less on par with “Windows,” then just close the ticket. As I mentioned earlier, you need to add various “cache” options during compilation, and this will speed up the build.

Yeah I think it is as fast as it's possible to be. If you try compiling in Debug, you will notice that the actual compilation is fast. Release mode uses very aggressive optimization (-O3) which will heavily increase compilation time but this is used by other projects as well. And like you said caching would improve this speed.
